packages/jupyter-ai/jupyter_ai/document_loaders/directory.py

```diff
@@ -1,54 +1,73 @@
-import logging
-from typing import Dict, List, Optional, Union
+from typing import List, Optional
+from pathlib import Path
+import hashlib
+import itertools
+
+import ray
 
 from langchain.document_loaders.base import BaseLoader
-from langchain.document_loaders.directory import FILE_LOADER_TYPE, _is_visible
-from langchain.document_loaders.unstructured import UnstructuredFileLoader
 from langchain.schema import Document
-from wcmatch.pathlib import Path
-
-
-logger = logging.getLogger(__name__)
+from langchain.document_loaders.directory import _is_visible
+from langchain.text_splitter import (
+    RecursiveCharacterTextSplitter, TextSplitter,
+)
 
+@ray.remote
+def path_to_doc(path):
+    with open(str(path), 'r') as f:
+        text = f.read()
+        m = hashlib.sha256()
+        m.update(text.encode('utf-8'))
+        metadata = {'path': str(path), 'sha256': m.digest(), 'extension': path.suffix}
+        return Document(page_content=text, metadata=metadata)
 
-class DirectoryLoader(BaseLoader):
-    """Loading logic for loading documents from a directory."""
+class ExcludePattern(Exception):
+    pass
+    
+def iter_paths(path, extensions, exclude):
+    for p in Path(path).rglob('*'):
```

Comment:
One last note, perhaps [`os.walk()`](https://docs.python.org/3/library/os.html#os.walk) could be used in lieu of walking the file tree imperatively.